File: bindings/go/ldp.go
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

// LDP protocol types for proven-servers.
// All tag values match the Idris2 ABI discriminants exactly.
package proven

// ContainerType represents the ContainerType type (Idris2 ABI tags).
type ContainerType uint8

const (
	ContainerTypeBasic ContainerType = iota
	ContainerTypeDirect
	ContainerTypeIndirect
)

// LdpResourceType represents the LdpResourceType type (Idris2 ABI tags).
type LdpResourceType uint8

const (
	LdpResourceTypeRdfSource LdpResourceType = iota
	LdpResourceTypeNonRdfSource
	LdpResourceTypeContainer
)

// Preference represents the Preference type (Idris2 ABI tags).
type Preference uint8

const (
	PreferenceMinimalContainer Preference = iota
	PreferenceIncludeContainment
	PreferenceIncludeMembership
	PreferenceOmitContainment
	PreferenceOmitMembership
)

// InteractionModel represents the InteractionModel type (Idris2 ABI tags).
type InteractionModel uint8

const (
	InteractionModelLdpr InteractionModel = iota
	InteractionModelLdpc
	InteractionModelLdpBasicContainer
	InteractionModelLdpDirectContainer
	InteractionModelLdpIndirectContainer
)

// ConstraintViolation represents the ConstraintViolation type (Idris2 ABI tags).
type ConstraintViolation uint8

const (
	ConstraintViolationMembershipConstant ConstraintViolation = iota
	ConstraintViolationContainsTriplesModified
	ConstraintViolationServerManaged
	ConstraintViolationTypeConflict
)

File: bindings/python/proven_servers/ldp.py
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
#
# Python bindings for the proven-ldp protocol types.

"""LDP protocol types for proven-servers."""

from enum import IntEnum


class ContainerType(IntEnum):
    """ContainerType matching the Idris2 ABI tags."""
    BASIC = 0
    DIRECT = 1
    INDIRECT = 2


class LdpResourceType(IntEnum):
    """LdpResourceType matching the Idris2 ABI tags."""
    RDF_SOURCE = 0
    NON_RDF_SOURCE = 1
    CONTAINER = 2


class Preference(IntEnum):
    """Preference matching the Idris2 ABI tags."""
    MINIMAL_CONTAINER = 0
    INCLUDE_CONTAINMENT = 1
    INCLUDE_MEMBERSHIP = 2
    OMIT_CONTAINMENT = 3
    OMIT_MEMBERSHIP = 4


class InteractionModel(IntEnum):
    """InteractionModel matching the Idris2 ABI tags."""
    LDPR = 0
    LDPC = 1
    LDP_BASIC_CONTAINER = 2
    LDP_DIRECT_CONTAINER = 3
    LDP_INDIRECT_CONTAINER = 4


class ConstraintViolation(IntEnum):
    """ConstraintViolation matching the Idris2 ABI tags."""
    MEMBERSHIP_CONSTANT = 0
    CONTAINS_TRIPLES_MODIFIED = 1
    SERVER_MANAGED = 2
    TYPE_CONFLICT = 3
